lib/rubygems/commands/rebuild_command.rb

```diff
@@ -0,0 +1,262 @@
+# frozen_string_literal: true
+
+require "date"
+require "digest"
+require "fileutils"
+require_relative "../package"
+
+class Gem::Commands::RebuildCommand < Gem::Command
+  DATE_FORMAT = "%Y-%m-%d %H:%M:%S.%N Z"
+
+  def initialize
+    super "rebuild", "Attempt to reproduce a build of a gem."
+
+    add_option "--force", "Skip validation of the spec." do |_value, options|
+      options[:force] = true
+    end
+
+    add_option "--strict", "Consider warnings as errors when validating the spec." do |_value, options|
+      options[:strict] = true
+    end
+
+    add_option "--source GEM_SOURCE", "Specify the source to download the gem from." do |value, options|
+      options[:source] = value
+    end
+
+    add_option "--original GEM_FILE", "Specify a local file to compare against (instead of downloading it)." do |value, options|
+      options[:original_gem_file] = value
+    end
+
+    add_option "--gemspec GEMSPEC_FILE", "Specify the name of the gemspec file." do |value, options|
+      options[:gemspec_file] = value
+    end
+
+    add_option "-C PATH", "Run as if gem build was started in <PATH> instead of the current working directory." do |value, options|
+      options[:build_path] = value
+    end
+  end
+
+  def arguments # :nodoc:
+    "GEM_NAME      gem name on gem server\n" \
+    "GEM_VERSION   gem version you are attempting to rebuild"
+  end
+
+  def description # :nodoc:
+    <<-EOF
+The rebuild command allows you to (attempt to) reproduce a build of a gem
+from a ruby gemspec.
+
+This command assumes the gemspec can be built with the `gem build` command.
+If you use either `gem build` or `rake build`/`rake release` to build/release
+a gem, it is a potential candidate.
+
+If the gem includes top-level files that change frequently (e.g. Gemfile.lock),
+it may require more effort to reproduce a build. For example, it might require
+more precisely matched versions of Ruby, RubyGems, or Bundler to be used.
```

Comment:
I think, since the running RubyGems version is [included in metadata](https://github.com/rubygems/rubygems/blob/a7aa64a918a590efe0e8c4da0eba189a8aa5140d/lib/rubygems/specification.rb#L153), I think an exact RubyGems version match is always necessary at the moment?

Reply:
yes -- one thing we could do is error out ahead of time if the wrong gem version is used, or add a way to override the rubygems version serialized into the gemspec, the same way we allow overriding the source date epoch?

Reply:
Yeah. I'd go with the latter, seems like it could make this more useful. Also I see some other differences in metadata, like some trailing spaces differences caused by the rubygems version divergence. We could also normalize those.

Reply:
I updated the help text to be more correct.